### PR TITLE
added `Morphism` definition to Function.Core

### PR DESCRIPTION
**Pull request by pnlph**
- added `Morphism` definition to Function.Core (de-deprecation)
- imported publicly in `Algebra.Morphism.Definitions` and `Relation.Binary.Morphism.Definitions`
- removed `Algebra.Core` import from `Relation.Binary.Morphism.Definitions`
- added Morphism to Changelog Version 1.4

### CODE REVIEW

File: CHANGELOG/v1.4.md [hunk 1]
```diff
@@ -12,15 +12,23 @@ Non-backwards compatible changes
...
+`map` now requires a function on both nodes and leaves. There are also
+`mapₙ` and `mapₗ` which only map on nodes and leaves respectively. There
+are similar changes for `Data.Tree.Relation.Unary.All`.
 
 Deprecated modules
 ------------------
 
+* In `Algebra.Morphism.Definitions` and `Relation.Binary.Morphism.Definitions`
+  the type `Morphism A B` were recovered by publicly importing its
+  definition from `Function.Core`. See discussion in issue #1206.
```
Comment (MatthewDaggitt) [inline CHANGELOG/v1.4.md:30]: This would more appropriately live under `Deprecated names` rather than `Deprecated modules`

**Comment by MatthewDaggitt** [comment]
If you look at the [test log](https://travis-ci.org/github/agda/agda-stdlib/jobs/689573729), it appears that it's clashing with `Morphism` elsewhere. Could you take a look a look and try and resolve it? You can run `make test` locally to check that it goes through correctly.

**Comment by pnlph** [comment]
Yes, the new definition clashes with the records `Morphism` in `Category.Functor` and `Morphism` in `Category.Applicative.Index`.

I renamed the first record to `CategoryMorphism` and the second one to `ApplicativeMorphism`; similarly to the way in which e.g. `Data.Container.Indexed` declares `ContainerMorphism`.

I tried to add a `WARNING_ON_USAGE` but it was clashing when defining:
```
Morphism = CategoryMorphism
{-# WARNING_ON_USAGE returnTC
"Warning: Morphism was deprecated in v1.4.
Please use CategoryMorphism instead."
#-}
```
Is there any other way of adding warnings?

**Comment by pnlph** [comment]
Still failing.

It seems like the records in `Algebra.Morphism` like `IsSemigroupMorphism`, `IsMonoidMorphism`, etc. fail to disambiguate `Morphism`, but they are actually pointing to the same definition:
```
  Algebra.Morphism.Morphism bound at
    /home/travis/build/agda/agda-stdlib/src/Function/Core.agda:27,1-9
  Function.Morphism bound at
    /home/travis/build/agda/agda-stdlib/src/Function/Core.agda:27,1-9
```

**Comment by pnlph** [comment]
I finally opted for removing the import in `Algebra.Morphism.Definitions`, I also hid `Morphism` when importing `Function` in `Algebra.Morphism`.

If it looks good to you, I will update the changelog accordingly.

File: src/Category/Applicative/Indexed.agda [hunk 2]
```diff
@@ -94,9 +94,10 @@ record RawIAlternative
 -- Applicative functor morphisms, specialised to propositional
 -- equality.
 
-record Morphism {I : Set i} {F₁ F₂ : IFun I f}
-                (A₁ : RawIApplicative F₁)
-                (A₂ : RawIApplicative F₂) : Set (i ⊔ suc f) where
+record ApplicativeMorphism
```
Comment (gallais) [inline src/Category/Applicative/Indexed.agda:97]: Can we keep these names & simply hide the `Morphism` import?

Comment (pnlph) [reply src/Category/Applicative/Indexed.agda:97]: Yes, you both know much better the intricacies of the `agda-stdlib` but please note that @MatthewDaggitt was [asking explicitly](https://github.com/agda/agda-stdlib/issues/1206#issuecomment-631807844) for the import.

I can change it to whatever you agree.

Comment (gallais) [reply src/Category/Applicative/Indexed.agda:97]: Re-exporting the definition from various modules is different from
hiding it when importing `Function.Core` inside of the `Category.*` modules.

I don't think it's worth introducing breaking changes for a simple type alias.

Comment (pnlph) [reply src/Category/Applicative/Indexed.agda]: I was hiding `Morphism` when importing `Function` in some modules, Travis is happy now :relieved:

**Comment by MatthewDaggitt** [comment]
Looks good to me. The only problem is that you've added the CHANGELOG entry to the wrong file, it should be in the top-level file `CHANGELOG.md`. It's my fault as I'm not quite sure why `CHANGELOG/v1.4md` exists... it shouldn't do until we release v1.4!

**Pull request merged**